### DISCUSSION TITLE
Improve FileSystem asset previews

### DISCUSSION THREAD

**Discussion by dallonf** [godotengine/godot-proposals#4837] [Editor]
In my most recent project, I imported some asset packs from Synty and Kenney. These were 3D assets, and browsing them looks something like this:

![image](https://user-images.githubusercontent.com/346300/178013315-47eb9b28-4631-4510-9fd4-64d55226d6cb.png)

Not very helpful! The only way to see what each of those models looks like is to drag it into the scene. I got so frustrated with this that I switched the project to Unity. Here's what the asset browser looks like there:

![image](https://user-images.githubusercontent.com/346300/178013825-5181cd7e-bbdf-40ee-971e-3799a8883480.png)

That's a little better - you can sort of see what they are. But even more importantly, when you click on one of them to select it, the Inspector panel opens and shows a _live preview_:

![image](https://user-images.githubusercontent.com/346300/178014068-1b3780e8-132e-4931-81de-b96c92627869.png)

You can rotate this preview to get a look from all angles. Oddly, you can't zoom and pan this preview, which would be even more helpful.

If Godot had something comparable to the second feature, I would have been able to continue working on this project in the engine.

**Comment by miv391** [comment]
In Godot 4 you can open the 3D model without adding it to a scene. You can also rotate and zoom.

![kuva](https://user-images.githubusercontent.com/49998025/178114591-cfcf12cb-b145-4932-9e85-01128f181a54.png)

**Reply by dallonf** [reply]
That helps!

1. Any chance that could be backported to Godot 3.x? 
2. How do you open this preview? I assume it's a double-click - if so, that's still a lot of friction compared to just a single-click and seeing it appear in a panel. It might sound like a nitpicky complaint, but considering you may be skimming through dozens to hundreds of assets, it'll add up fast.

**Reply by miv391** [reply]
The window (which is titled Advanced Import Setting) is opened by double-clicking. To open another model you need to close the window first. So this is not an optimal tool for browsing the models. 

Proposal: It would be nice you could open next/previous model for example by pressing page up/down keys. Or if the Advanced Import Setting would open in the main work area as tab page instead of separate window, you could just double-click another model without closing the current model first.

**Reply by Calinou** [reply]
> Any chance that could be backported to Godot 3.x?

The 3D import system was rewritten from scratch for 4.0, and it's backwards-incompatible with 3.x.

**Comment by xix-xeaon** [comment]
Previews seem to work well for obj-files, can't basically the same be done for fbx and gltf?

Shaders don't have previews, only materials do. Shaders should be previewed similar to materials just with default parameters. The icon differentiates shaders from materials but there could be some other difference in the preview as well to make it clearer.

**Scene previews** are very important but really weird and currently only better than nothing. It seems the preview is generated from a center zoomed in version of your editor camera when you save the scene. This makes the previews vary wildly and I don't think it makes much sense - the preview should show the scene as a whole, not what you were last looking at.

For GUI scenes it should simply be a preview of the whole GUI. For 3D scenes it should use the AABB and preview similar to how it's done with obj-files - it uses a bit of angle and I think it's orthogonal, or close to it. Also, it currently includes all gizmos in the preview as well. It definitely shouldn't include "current selection"-gizmos, like the tool for moving things around, and probably no gizmos at all (it just makes the small preview cluttered). For 2D it should preview the whole content like with GUI. For mixed scenes it should just mix the contents the same as a camera in game normally would.

The current scene-preview behavior could be kept as a setting but I don't think it should be default. Also I think I'd want backgrounds to be transparent in previews, but this might be a setting as well. For 3D including the entire AABB might make the actual scene contents very small, a zoom setting could be useful.

Edit: Also, the split-mode which gives you those previews takes up too much space in either of the side docks. The file/preview part of the split should probably be a panel in the middle bottom (where output is). (Selecting, or even hovering, could use space in a side-dock for a large preview if it was temporary.) I believe there's a bit more flexibility for docks in Godot 4 but I don't think you can move those splits individually.

**Reply by Calinou** [reply]
See https://github.com/godotengine/godot-proposals/issues/4759 for improving editor scene preview generation. The proposed solution requires a bit of manual work, but it can give you *consistently* useful previews where automatic solutions can fail.

> Previews seem to work well for obj-files, can't basically the same be done for fbx and gltf?

By default, OBJ files are 3D *meshes* while FBX and glTF are 3D *scenes*.[^1] Godot can't create scene previews of entire scenes before loading and saving them, and since you can't save FBX and glTF scenes from Godot, there is no chance of creating a preview for them.

[^1]: You can change the import mode of .obj to a 3D scene, but you can't change the import mode of a FBX or glTF to a 3D mesh.

**Reply by azur-wolve** [reply]
For me at least, a good thing is to be able to capture an image from either the Viewport (or also form a camera preview in 3D workspace).
Instead of have a whole image that is a mess of all the contents, let the user to capture a distinctive element/angle.

It can become a bit of work ye, but its better than nothing, and also works well for both pre-fabs and even in-game in-game representations/icons for objects.
Mount & Blade: Warband does something like that, but it does automatically, based on a fixed angle relative to the forward axis.
This can be custom implemented,  but have it built-in with such flexibility sounds really useful imo.

Also useful for better dev-team communication (assets, scenes,..).

**Reply by xix-xeaon** [reply]
It makes no difference that they're technically *imported* as scenes. I need to see the assets when I browse them, before even choosing which one to import. Also, a mesh needs to be loaded too before Godot can generate a thumbnail. Just make the thumbnail with the default import settings until I change them, then you can remake the thumbnail based on the new settings. It's not rocket science.

I made a lengthy response on that proposal regarding scene previews.

**Comment by azur-wolve** [comment]
Something to consider:
Thumbnails would add loading overhead on import, and need to be updated when assets get modified (re-import)

**Reply by xix-xeaon** [reply]
Even if it took significant time it would still be worth it. However, previews for OBJ files are *not* generated on import, they're generated when you browse them in the FileSystem dock. It takes a short while if you have many files but it's done in a separate thread or something because the editor is still completely responsive. It should be the same for the other file types.

**Comment by dallonf** [comment]
I'm less interested in thumbnails, and more interested in an preview panel that could appear just by highlighting an asset.

**Comment by dallonf** [comment]
I made a prototype of what this could look like as an editor plugin: 

https://user-images.githubusercontent.com/346300/180298567-5f63de76-9a92-4fdb-ae87-f76a69b57eb6.mp4

Here's the source: https://github.com/dallonf/GodotAssetPreviewPlugin

I'd love it if this were just part of the FileSystem dock, though, at least optionally. Do you think the Godot team would accept a PR to add a feature like this?

**Reply by Calinou** [reply]
> I'd love it if this were just part of the FileSystem dock, though, at least optionally. Do you think the Godot team would accept a PR to add a feature like this?

Improved FileSystem dock tooltips are currently being worked on in https://github.com/godotengine/godot/pull/63263 (without interactive 3D previews). I understand you don't want this to only appear on hover, but it's the least intrusive solution we have right now.

The issue with the proposed solution here is that it takes a lot of vertical space (which is precious on today's 16:9 displays). Even if you made the preview resizable and toggleable, it may still hamper usability on 1366×768 laptops which are still fairly common among Godot users. (The preview shouldn't be disabled by default to prevent harming the feature's discoverability.)

**Reply by dallonf** [reply]
Perhaps the preview could be a separate dock?

**Reply by Calinou** [reply]
You can only display 3 docks at a time by default[^1], so I don't know if any of the current dock locations would be suited for this. If people have to switch away from the Scene/Import dock or inspector to preview assets, it's not exactly helpful from a workflow perspective.

[^1]: You can display up to 8 docks at a time with custom layouts, but it's not viable on most screen resolutions.

**Reply by dallonf** [reply]
Oh, another thing that would improve this situation is if the `FileSystemDock` could trigger signals when its selection changes. That way you could make an editor plugin, if you want, that shows an interactive preview in another dock or the bottom panel.

**Reply by Zireael07** [reply]
More on docks: https://github.com/godotengine/godot-proposals/issues/4565

Your proposal would probably be received better AFTER the current problems are fixed

**Reply by miv391** [reply]
If you need to save screen space, how about displaying either the asset list or the asset preview? Start by displaying the asset list. When an asset is clicked, change to the preview. Display the 3D object, its name, position on list (e.g. "5/42") and close button. In this preview allow changing the selected asset using keyboard (cursor keys, page up/down, home/end) and closing the preview with ESC key. Now you could iterate through the assets using keyboard.

**Reply by dallonf** [reply]
That sounds like a very particular kind of UI, frankly, so it'd be better off as a plugin IMO. I'm increasingly in the plugin camp, at least until some of the general UI concerns are taken care of. (although if one plugin becomes very popular, it would make sense to adopt it into the core) The engine would just need a couple extra hooks to make this kind of plugin easier.

**Comment by Bat-Computer** [comment]
As of writting this we're on version 4.2.1 Stable, it's nearly 2024 and there's been no update on adding thumbnail previews for imported 3D objects in Godot. I don't understand why it's not a big priority. Other games engines such as Unity and Unreal have had thumbnail previews for 3D models for over a decade or much longer. Godot 4+ has been striving for more 3D workflows and better renderes, but with all the new features some core fundimental features such as 3D thumbnail previews haven't been addressed, I really don't understand why. The only reason I can think currently is that the user base currenlty are more into 2D games and thus haven't raised this issue enough.

Currently there is a a paid plugin called 'Asset Placer' by Cookie Badger (https://cookiebadger.itch.io/assetplacer), however this plugin costs $18. Although it adds many extra features as well as displaying thumbnails for 3D models - one caviet though is that it is for the Mono version of Godot only (as it's written is C#) and with plugins like these there's the problem of it breaking in the future if it's not maintained. I don't want to rely on plugins that add such import functionality, i've been bitten with Unity over the years the same way.

I really wish this could be addressed, not just for me, but for other's to take Godot more seriously. As the oringal poster discussed, like him I have many 3D assets packs that are nearly impossible to use as many packs have 300+ models, none of which I can see in the editor, without having to double-click on each 3D model to launch the 'Advanced Import Settings' window. This is very time-consuming and makes it very very slow to create levels.

I could of course build my levels in Blender, but when muliple assets are used, such as walls, roads, light posts, fences etc, the filesize is huge! From my humble knoweldge, professional games are made with modular assets and laid out typically inside the game engine itself. That way you only have to have one copy of each modular asset to reuse them and layout the level.

Making the switch to Godot is difficult when this hasn't been addressed for me and I'm sure other's feel the same, it's a great engine when it comes to 2D, but usability/workflow wise with 3D, it's very laborious when it doesn't need to be. If I knew how to code a plugin myself to add the thumbnail functionality, I would, but sadly I don't even know where to start. But if I could, it would be free and not $18.

Hats off to Cookie Badger for his plugin and the time and effort he put into it, but we really need this feature added, I'd even pay something for it to be written and included in Godot by default! ;)

**Comment by Calinou** [comment]
- Closing in favor of https://github.com/godotengine/godot-proposals/issues/8760.
